### PR TITLE
Adds `no-bad-enzyme-features` rule

### PR DESCRIPTION
**Pull request by cartogram**
Part of https://github.com/Shopify/eslint-plugin-shopify/issues/77

#### Questions
1. What are all the [bad parts](https://github.com/Shopify/eslint-plugin-shopify/pull/163/files#diff-82854c898309f636ecdc9af8dded2ae2R71) we want to prevent? 
1. Should this be configurable? probably.
1. better name ideas?

Will follow up with the docs, changelog, readme, etc...

### CODE REVIEW

File: docs/rules/enzyme-no-bad-features.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Restricts usage of enzyme to the good parts. (enzyme-no-bad-features)
```
Comment (lemonmade) [inline docs/rules/enzyme-no-bad-features.md]: I feel this name is appropriate tbh

File: lib/rules/enzyme-no-bad-features.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    return {
+      'Program:exit': function() {
+        enzymeWrappers.forEach(checkNodeForInvalidProps);
+      },
+      CallExpression(node) {
+        if (!isEnzymeWrapper(node)) {
+          return;
+        }
+
+        enzymeWrappers.push(node);
```
Comment (lemonmade) [inline lib/rules/enzyme-no-bad-features.js]: I don’t see the benefit of pushing to an array and doing it all on exit. What is the difference from just doing it at the AST node itself?

File: lib/rules/enzyme-no-bad-features.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      },
+    };
+
+    function report(node, prop) {
```
Comment (lemonmade) [inline lib/rules/enzyme-no-bad-features.js]: I know this works but IMO air is very confusing to see things after an unguarded return.

File: lib/rules/enzyme-no-bad-features.js (hunk 2 above)
Comment (lemonmade) [reply lib/rules/enzyme-no-bad-features.js]: Never mind I see it now

File: lib/rules/enzyme-no-bad-features.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+      const variableDeclaratorParent = findParent(
+        node,
+        (parent) => parent.type === 'VariableDeclarator',
+      );
+
+      if (variableDeclaratorParent) {
+        const references = context
+          .getDeclaredVariables(variableDeclaratorParent)[0]
+          .references.slice(1);
```
Comment (lemonmade) [inline lib/rules/enzyme-no-bad-features.js]: Why the slice?

File: lib/rules/enzyme-no-bad-features.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  return [
+    'setState',
+    'setContext',
+    'setProps',
+    'instance',
+    'state',
+    'simulate',
+    'simulateError',
+    'ref',
+    'context',
```
Comment (lemonmade) [inline lib/rules/enzyme-no-bad-features.js]: We still need them to get context, and to set props, in the course of normal tests, so IMO not the right defaults. Agree it should be configurable though.

**Review by lemonmade — CHANGES_REQUESTED** [review]

**Review by GoodForOneFare — COMMENTED** [review]
I'm oversensitive to computers yelling at me, so messages like "No bad enzyme features" gets my dander up.

Splitting into at least two rules would provide better context for error messages:
- `shopify/enzyme/immutable` (forbids set*, simulate*)
- `shopify/enzyme/component-boundaries` (forbids instance, state, ref, context)

Not keen on the second rule's naming, but you get the idea 😅

**Comment by cartogram** [comment]
@lemonmade this is ready for a review. No rush. Also if you spot what piece of config I'm missing to pass the `all` test, please let me know 🙏

File: docs/rules/enzyme/no-bad-features.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+You can learn more about each of these features on the [Enzyme documentation](https://airbnb.io/enzyme).
+
+Examples of **incorrect** code for the `immutability` group of features:
+
+```ts
+const wrapper = mount(<Foo />);
+wrapper.setState({foo: 'bar'});
+wrapper.setContext({bar: 'baz'});
+wrapper.simulate('click');
+wrapper.simulateError(new Error('Explosions!'));
```
Comment (lemonmade) [inline docs/rules/enzyme/no-bad-features.md:29]: I don’t know that these two fit as cleanl into immutability. IMO, they are kind of their own class (maybe separate rule?), and the other two here fit in to component boundaries

File: lib/rules/enzyme/no-bad-features.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  create(context) {
+    const enzymeWrappers = [];
+    const options = context.options[0] || {};
+    const allowed = options.allow || [];
+
+    function report(node, prop) {
+      context.report({
+        node,
+        message:
+          'No bad enzyme features. Avoid using {{prop}} on an enzyme wrapper.',
```
Comment (lemonmade) [inline lib/rules/enzyme/no-bad-features.js]: I think we might need or-feature recommendations in place of “don’t do bad”

File: lib/rules/enzyme/no-bad-features.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        node,
+        (parent) => parent.type === 'VariableDeclarator',
+      );
+
+      if (variableDeclaratorParent) {
+        const references = context
+          .getDeclaredVariables(variableDeclaratorParent)[0]
+          .references.slice(1);
+        references.forEach((reference) =>
+          checkNodeForInvalidProps(reference.identifier),
```
Comment (lemonmade) [inline lib/rules/enzyme/no-bad-features.js:80]: Seems odd that this is basically recursive?

**Review by lemonmade — CHANGES_REQUESTED** [review]

File: lib/rules/enzyme/no-bad-features.js (hunk 8 above)
Comment (cartogram) [reply lib/rules/enzyme/no-bad-features.js:80]: How is it odd? Basically want to unwarp the node and check the references for invalid props.

File: lib/rules/enzyme/no-bad-features.js (hunk 7 above)
Comment (cartogram) [reply lib/rules/enzyme/no-bad-features.js]: Yeah good call.

**Comment by cartogram** [comment]
As we move away from enzyme, I don't see a need for this rule anymore.

**Pull request closed without merging**